### ISSUE TITLE
DispNetCorr1D always gives negative values 

### ISSUE THREAD

**Issue by adhara123007** [lmb-freiburg/dispnet-flownet-docker#14]
Hi,

This is regarding the DispNet network. I am using the DispNetCorr1D network to register images. I understand the network is only trained to register Right image to the left image. Thus the DispNetCorr1D always gives negative values. In my case, I want to get both positive and negative values of shift. How can I use the weights of the current network to register both Left to Right images and Right to Left images?
For example if I just do
Left to Right registration
python run-flownet-dispnet.py ../models/DispNetCorr1D/model/DispNetCorr1D_CVPR2016.caffemodel ../models/DispNetCorr1D/model/deploy.prototxt ../data/dispnet/imgL.png ../data/dispnet/imgR.png ../data/dispnet/floLR.flo

![image](https://user-images.githubusercontent.com/40571095/55615518-8e3bdc80-5787-11e9-8cf2-ff5fe80e6a8b.png)

If I do Right to Left registration

python run-flownet-dispnet.py ../models/DispNetCorr1D/model/DispNetCorr1D_CVPR2016.caffemodel ../models/DispNetCorr1D/model/deploy.prototxt ../data/dispnet/imgR.png ../data/dispnet/imgL.png ../data/dispnet/floRL.flo

![image](https://user-images.githubusercontent.com/40571095/55615548-9bf16200-5787-11e9-98d9-5300d5dbf823.png)

In the second case, the disparity map should have been positive.

**Comment by nikolausmayer** [comment]
- you have to adjust the correlation layer to allow for correlation in both left and right directions (it's a parameter)
- depending on the network, there may be rectification layers that guarantee that the output has a unified sign
- and you probably have to retrain the whole thing :wink:

**Comment by adhara123007** [comment]
Thanks Nikolaus.

I didn't modify the network. But I tried training the network with my examples. I see the loss doesn't change between different resolutions. Is this the way it should look like?

![image](https://user-images.githubusercontent.com/40571095/55725229-285c8880-5a05-11e9-9cf5-bb61aef18ee1.png)

**Comment by adhara123007** [comment]
I am using the DispNetCorr1D layer. The correlation layer is by default both sided (single_direction = 0) . So, I guess no problem there. 
What type of rectification layers should I look at?

**Comment by nikolausmayer** [comment]
> What type of rectification layers should I look at?

ReLU-style layers :wink: 

I'm not sure if it is good that your loss is so similar across resolutions. Unfortunately I don't have the resources to investigate this at the moment.

**Comment by ADHARACSM** [comment]
Thank Nikolaus.
I fine tuned the network on my grayscale geophysics images and it works now. I also removed your NegRELU layers and added few layers in between.
Currently, I scale my images between 0 and 255, which might be inappropriate for my case ( the range in my case could be between -1000 to 1000) . Also, I would like to input images as float. So, I also write the images as .flo files and then create lmdb database from them So, I read the images as UNIT16FLO while training.
However, the network doesn't converge. What do you think should I tweak ?
I started with the learning rate (finetuning your network) and the network either overshoots or doesn't update after first few iterations.

**Comment by nikolausmayer** [comment]
@ADHARACSM 
- Did you confirm that the network can properly read your UINT16FLO data as images? My intuition would be that a lot of things could go wrong there.
- Our augmentation layers probably only work on image data with range [0,1]. Did you check for our rescaling layers before augmentation, and that what comes out of augmentation is still ok?

**Comment by adhara123007** [comment]
Thanks Nikolaus for the relevant questions.
I rechecked. The network reads UNIT16FLO data as images correctly. I believe the custom data layer divides by 32 when the data type is UINT16FLO. So, I multiplied by 32 when creating flo.
I had scaled the images between (-255) to (255). And Eltwise layer scales the images from (-1) to (1).
As for the Data Augmentation layer, I donot need the augmentation. So, I changed it.
Can you please help me how does recompute mean work? Does it randomly picks a sample 1000 times, computes mean 1000 times and takes the average?

layer {
  name: "img0s_aug_change"
  type: "DataAugmentation"
  bottom: "blob3"
  top: "img0_aug"
  top: "blob6"
  propagate_down: false 
  augmentation_param {
    max_multiplier: 1
    augment_during_test: false
    recompute_mean: 1000
    mean_per_pixel: false
    crop_width: 768
    crop_height: 384 
   }
}

**Comment by nikolausmayer** [comment]
I believe `recompute_mean` computes an average over the first 1000 samples it sees, then freezes that value. It was our attempt to get a meaningful dataset mean for both train and test settings, while avoiding dragging out this computation indefinitely.

**Comment by nikolausmayer** [comment]
(closed due to inactivity)
